src/go/rdctl/cmd/set.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright © 2022 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/spf13/cobra"
+	"os"
+	"strings"
+)
+
+type serverSettings struct {
+	Kubernetes struct {
+		ContainerEngine string `json:"containerEngine"`
+		Enabled         bool   `json:"enabled"`
+		Version         string `json:"version"`
+	} `json:"kubernetes,omitempty"`
+}
+
+var specifiedSettings serverSettings
+
+// setCmd represents the set command
+var setCmd = &cobra.Command{
+	Use:   "set",
+	Short: "Update selected fields in the Rancher Desktop UI and restart the backend.",
+	Long: `The following options are supported:
+    --container-engine=containerd|moby
+    --kubernetes-enabled=true|false
+    --kubernetes-version=VERSION
+
+The '=' sign can be replaced by one or more spaces.
+'docker' is an accepted synonym for 'moby'.
+`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) > 0 {
+			return fmt.Errorf("set command: unrecognized command-line arguments specified: %v", args)
+		}
+		return doSetCommand()
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(setCmd)
+	setCmd.Flags().StringVar(&specifiedSettings.Kubernetes.ContainerEngine, "container-engine", "", "Set engine to containerd or moby (aka docker).")
+	setCmd.Flags().BoolVar(&specifiedSettings.Kubernetes.Enabled, "kubernetes-enabled", false, "Control whether kubernetes runs in the backend.")
+	setCmd.Flags().StringVar(&specifiedSettings.Kubernetes.Version, "kubernetes-version", "", "Choose which version of kubernetes to run.")
+}
+
+/**
+ * Get the current settings as a JSON string,
+ * and unmarshal the string into a settings block that we care about.
+ * The update only the fields in that block which are specified on the command-line.
+ * Send that block back to the server in a PUT set command.
+ */
+func doSetCommand() error {
+	result, err := doRequest("GET", "list-settings")
+	if err != nil {
+		return err
+	}
+	var currentSettings serverSettings
+	err = json.Unmarshal(result, &currentSettings)
+	if err != nil {
+		return err
+	}
```

Comment:
Why do we need fetch the current settings? I thought that posting a _partial_ settings will only modify the affected bits.

Reply:
In order to distinguish `set --kubernetes-enabled=false ...` from `set <not mentioning kubernetes-enabled at all>`. If you say that's the same as `--kubernetes-enabled=false` , and it's currently enabled, you're violating the principle of least surprise.

Reply:
@ericpromislow I think @mook-as is saying that if the user didn't specify `--kubernetes-enabled`, then the client should not include this property at all in the request to the server.

The server would only modify the specified settings.

So I think you are in agreement on how it should work from the user's point of view, you just make different assumptions about the implementation.

Reply:
Then I have to build the JSON payload by hand. Which is what I originally did, but it seemed clunky. I thought the current form is more elegant, even if it does require a second call to the server.

It sounds like both of you would prefer I build the JSON payload the first way.

Reply:
For this release I want you to do whatever is faster to implement; we are revisiting this for `v1` anyways.

You don't have to build the payload by hand if you use pointers, so the value can be `nil`. I assume this would also work on the server side.

But I think for now it will be easiest to keep the additional API call.

Reply:
All solved by using indirect objects and `omitempty` in JSON blocks.